File: telemetry-cpp/examples/metrics.cpp
#include "telemetry/telemetry.hpp"

#include <chrono>
#include <thread>
#include <vector>
#include <cstdlib>

std::pair<std::string, uint16_t> get_otel_endpoint() {
    const char* env = std::getenv("OTEL_EXPORTER_OTLP_ENDPOINT");
    if (env) {
        std::string endpoint(env);
        auto pos = endpoint.find(':');
        if (pos != std::string::npos) {
            return {endpoint.substr(0, pos),
                    static_cast<uint16_t>(std::stoi(endpoint.substr(pos + 1)))};
        }
        return {endpoint, 4317};
    }
    return {"localhost", 4317};
}

struct LlmMetrics : public telemetry::metrics::RecordMetrics {
    uint64_t request_count = 0;
    double latency_ms = 0.0;
    double cache_hit_rate = 0.0;

    LlmMetrics() = default;
    LlmMetrics(uint64_t count, double latency, double hit_rate)
        : request_count(count), latency_ms(latency), cache_hit_rate(hit_rate) {}

    std::vector<telemetry::metrics::MetricField> metric_fields() const override {
        return {
            {"llm.requests.total", telemetry::metrics::MetricType::Counter,
             "Total number of LLM requests", static_cast<double>(request_count)},
            {"llm.response.latency_ms", telemetry::metrics::MetricType::Histogram,
             "LLM response latency in milliseconds", latency_ms},
            {"llm.cache.hit_rate", telemetry::metrics::MetricType::Gauge,
             "LLM cache hit rate percentage", cache_hit_rate}
        };
    }
};

int main() {
    auto [otel_host, otel_port] = get_otel_endpoint();

    auto telemetry = telemetry::Telemetry::builder("metrics-example", "1.0.0")
        .description("Metrics example service")
        .environment(telemetry::Environment::Development)
        .with_mcap("examples/metrics.mcap")
        .with_otlp(otel_host, otel_port)
        .build();

    TELEMETRY_LOG_INFO("Metrics Example Started");
    TELEMETRY_LOG_INFO("Sending metrics to OTEL collector at localhost:4317");

    LlmMetrics llm_metrics{42, 123.5, 0.85};
    telemetry.metrics().record(llm_metrics);
    TELEMETRY_LOG_INFO("Recorded LLM metrics from struct");

    TELEMETRY_LOG_INFO("Recording metrics for 30 seconds...");

    for (int iteration = 0; iteration < 30; ++iteration) {
        for (int i = 0; i < 10; ++i) {
            telemetry.metrics().counter("api.requests", 1.0);
            telemetry.metrics().histogram("api.latency_ms", static_cast<double>(i) * 10.0 + 50.0);
            telemetry.metrics().gauge("api.active_connections", static_cast<double>(10 - i));

            telemetry.metrics().record(llm_metrics);

            TELEMETRY_LOG_DEBUG("Recorded API metrics");
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }
    }

    TELEMETRY_LOG_INFO("Metrics recording completed");
    TELEMETRY_LOG_INFO("MCAP file will be finalized automatically");

    return 0;
}

File: telemetry-rs/telemetry/examples/metrics.rs
use telemetry::derive::Metrics;
use telemetry::{Environment, Telemetry, logger};

#[derive(Debug, Metrics)]
pub struct LlmMetrics {
    #[metric(
        name = "llm.requests.total",
        description = "Total number of LLM requests",
        counter
    )]
    pub request_count: u64,

    #[metric(
        name = "llm.response.latency_ms",
        description = "LLM response latency in milliseconds",
        histogram
    )]
    pub latency_ms: f64,

    #[metric(
        name = "llm.cache.hit_rate",
        description = "LLM cache hit rate percentage",
        gauge
    )]
    pub cache_hit_rate: f64,
}

/// Runs the metrics example, recording sample LLM and API telemetry.
///
/// # Examples
///
/// Run the example binary to start recording metrics:
///
/// ```text
/// cargo run --example metrics
/// ```
///
/// # Errors
///
/// Returns an error if telemetry initialization or metric recording fails.
#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // Auto-discovers telemetry.toml config file
    let mut telemetry = Telemetry::new()
        .with_service("metrics-example", "1.0.0")
        .description("Metrics example service")
        .environment(Environment::Development)
        .with_mcap("telemetry/examples/metrics.mcap")
        .build()?;

    logger::info!("Metrics Example Started");
    logger::info!("Sending metrics to OTEL collector at localhost:4317");

    // Record metrics using the macro-generated struct
    let llm_metrics = LlmMetrics {
        request_count: 42,
        latency_ms: 123.5,
        cache_hit_rate: 0.85,
    };

    telemetry.metrics.record(&llm_metrics)?;
    logger::info!("Recorded LLM metrics from struct");

    // Or record metrics directly
    logger::info!("Recording metrics for 30 seconds...");

    for _iteration in 0..30 {
        for i in 0..10 {
            telemetry.metrics.counter("api.requests", 1.0)?;
            telemetry
                .metrics
                .histogram("api.latency_ms", (i as f64) * 10.0 + 50.0)?;
            telemetry
                .metrics
                .gauge("api.active_connections", (10 - i) as f64)?;

            // Re-record LLM metrics every iteration to keep them visible
            telemetry.metrics.record(&llm_metrics)?;

            logger::debug!("Recorded API metrics");
            tokio::time::sleep(tokio::time::Duration::from_millis(100)).await;
        }
    }

    logger::info!("Metrics recording completed");
    logger::info!("MCAP file will be finalized automatically");

    Ok(())
}
